src/Types/TxConstraints.purs

```diff
@@ -0,0 +1,370 @@
+module Types.TxConstraints
+  ( InputConstraint(..)
+  , OutputConstraint(..)
+  , TxConstraint(..)
+  , TxConstraints(..)
+  , addTxIn
+  , isSatisfiable
+  , modifiesUtxoSet
+  , mustBeSignedBy
+  , mustHashDatum
+  , mustIncludeDatum
+  , mustMintCurrency
+  , mustMintCurrencyWithRedeemer
+  , mustMintValue
+  , mustMintValueWithRedeemer
+  , mustPayToOtherScript
+  , mustPayToPubKey
+  , mustPayToTheScript
+  , mustProduceAtLeast
+  , mustProduceAtLeastTotal
+  , mustSatisfyAnyOf
+  , mustSpendAtLeast
+  , mustSpendAtLeastTotal
+  , mustSpendPubKeyOutput
+  , mustSpendScriptOutput
+  , mustValidateIn
+  , pubKeyPayments
+  , requiredDatums
+  , requiredMonetaryPolicies
+  , requiredSignatories
+  , singleton
+  ) where
+
+import Prelude hiding (join)
+
+import Data.Array as Array
+import Data.Array ((:))
+import Data.Bifunctor (class Bifunctor)
+import Data.BigInt (BigInt)
+import Data.Foldable (class Foldable, any, foldl, foldMap, foldr, null)
+import Data.Generic.Rep (class Generic)
+import Data.Lattice (join)
+import Data.Map (fromFoldableWith, toUnfoldable)
+import Data.Maybe (Maybe(Just, Nothing))
+import Data.Newtype (class Newtype, over, unwrap)
+import Data.Show.Generic (genericShow)
+import Data.Tuple (uncurry)
+import Data.Tuple.Nested ((/\), type (/\))
+import Types.PlutusData
+  ( class ToData
+  , Datum(Datum)
+  , DatumHash
+  , toData
+  , unitRedeemer
+  )
+import Types.Interval (POSIXTimeRange, always, intersection, isEmpty)
+import Types.RedeemerTag (RedeemerTag(Mint))
+import Types.ScriptHash (MintingPolicyHash, ValidatorHash)
+import Types.Transaction (Redeemer)
+import Types.UnbalancedTransaction (PubKeyHash, TxOutRef)
+import Types.Value (TokenName, Value, currencyMPSHash, getNonAdaAsset, isZero)
+
+--------------------------------------------------------------------------------
+-- TxConstraints Type and related
+--------------------------------------------------------------------------------
+-- Taken from https://playground.plutus.iohkdev.io/doc/haddock/plutus-ledger-constraints/html/Ledger-Constraints.html
+
+-- | Constraints on transactions that want to spend script outputs
+data TxConstraint
+  = MustIncludeDatum Datum
+  | MustValidateIn POSIXTimeRange
+  | MustBeSignedBy PubKeyHash
+  | MustSpendAtLeast Value
+  | MustProduceAtLeast Value
+  | MustSpendPubKeyOutput TxOutRef
+  | MustSpendScriptOutput TxOutRef Redeemer
+  | MustMintValue MintingPolicyHash Redeemer TokenName BigInt
+  | MustPayToPubKey PubKeyHash Value
+  | MustPayToOtherScript ValidatorHash Datum Value
+  | MustHashDatum DatumHash Datum
+  | MustSatisfyAnyOf (Array TxConstraint)
```

Comment:
I'm curious why is it a 2D-list in plutus, but 1D here?

Reply:
Hang on, it looks like it's 1D on the latest version on `master` which what I was following, but `staging` is 2D (or whatever they use for https://playground.plutus.iohkdev.io/doc/haddock/index.html), maybe that is what confused me. https://github.com/input-output-hk/plutus/tree/ead3960ad6f70a7b3f88bcc1bc53eca756746f2a I am using this and running haddock on localhost.

Reply:
![Screenshot from 2022-02-24 08-15-35](https://user-images.githubusercontent.com/81221015/155485069-848c2e3b-1a20-4ddf-a52d-c1c6a6342a5d.png)

Reply:
Oh wait, I think I may be using an out of date build. Perhaps it is 2d after all.

Reply:
Okay, turns out I need to update `MustPayToPubKey` to `MustPayToPubKeyAddress` too.